### ISSUE TITLE
Align allowed package names between build and dependencies.

### ISSUE THREAD

**Issue by bame-da** [digital-asset/daml#7317]
I can compile a project with name `1_Foo`, but not use it as a data-dependency. I get

```
ghc-pkg: 0: Parse of field '
unexpected all digits in portion of unqualified component name

MyProject
' failed.
damlc: callProcess: /Users/bernhardelsner/.daml/sdk/1.5.0-snapshot.20200902.5118.0.2b3cf1b3/damlc/resources/ghc-pkg "recache" "--global-package-db=.daml/package-database/1.8/package.conf.d" "--expand-pkgroot" (exit 1): failed
```

We should either restrict project names, or fix dependencies with such names. The latter would be better as projects with `_` in their name may already exist "out there", but either is fine.

**Comment by sofiafaro-da** [comment]
Similar to #7208!

**Comment by hurryabit** [comment]
@robin-da And that's the other one. Could you please look into this once you find time for it? Thank you.

**Comment by robin-da** [comment]
@hurryabit , @bame-da This is a limitation by `ghc-pkg`. Valid package id's for `ghc-pkg` are specified by : https://downloads.haskell.org/~ghc/latest/docs/html/libraries/Cabal-3.2.0.0/Distribution-Types-PackageId.html. In particular, `_` is not allowed in package names. My suggestion is to disallow underscores as well since we don't want to patch `ghc-pkg`. This would however be a breaking change in DAML-LF. Thus, we could simply throw an error when parsing the package id from the daml.yaml and leave DAML-LF untouched.

**Comment by bame-da** [comment]
Could we make it a warning? Something like "Package names/versions should have format [RegExp]. You may be able to compile packages with different formats, but you will not be able to use them as dependencies in other projects. Unsupported package names or versions may start causing compilation errors without warning."

**Comment by hurryabit** [comment]
I like what @bame-da suggested. Let's do this. It's similar in style to what we did with regarding language extensions that might not work with data-dependencies.

**Comment by robin-da** [comment]
Fixed with #8254 .
